src/main/java/software/amazon/event/kafkaconnector/offloading/S3EventBridgeEventDetailValueOffloading.java

```diff
@@ -0,0 +1,172 @@
+import software.amazon.awssdk.services.eventbridge.model.PutEventsRequestEntry;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.event.kafkaconnector.EventBridgeResult;
+import software.amazon.event.kafkaconnector.logging.ContextAwareLoggerFactory;
+import software.amazon.event.kafkaconnector.util.MappedSinkRecord;
+
+public class S3EventBridgeEventDetailValueOffloading
+    implements EventBridgeEventDetailValueOffloadingStrategy {
+
+  private static final Logger logger =
+      ContextAwareLoggerFactory.getLogger(S3EventBridgeEventDetailValueOffloading.class);
+
+  private final String bucketName;
+  private final S3Client client;
+  private final ReplaceWithDataRefJsonTransformer detailValueTransformer;
+  private final JsonConverter jsonConverter;
+
+  public S3EventBridgeEventDetailValueOffloading(
+      final S3Client client, final String bucketName, final String jsonPathExp) {
+    this.bucketName = bucketName;
+    this.client = client;
+    this.detailValueTransformer = replaceWithDataRef(jsonPathExp);
+    this.jsonConverter = new JsonConverter();
+    this.jsonConverter.configure(singletonMap("schemas.enable", "false"), false);
+  }
+
+  @Override
+  public EventBridgeEventDetailValueOffloadingResult apply(
+      List<MappedSinkRecord<PutEventsRequestEntry>> putEventsRequestEntries) {
+
+    final Map<Boolean, List<EventBridgeResult<PutEventsRequestEntry>>> result =
+        putEventsRequestEntries.stream()
+            .map(this::apply)
+            .collect(partitioningBy(EventBridgeResult::isSuccess));
+
+    var success = result.get(true).stream().map(EventBridgeResult::success).collect(toList());
+    var errors = result.get(false).stream().map(EventBridgeResult::failure).collect(toList());
+
+    return new EventBridgeEventDetailValueOffloadingResult(success, errors);
+  }
+
+  private EventBridgeResult<PutEventsRequestEntry> apply(
+      MappedSinkRecord<PutEventsRequestEntry> mappedSinkRecord) {
+    try {
+      var s3key = generateS3Key(mappedSinkRecord.getSinkRecord());
+      var putEventsRequestEntry =
+          mappedSinkRecord
+              .getValue()
+              .copy(
+                  it ->
+                      it.detail(
+                          detailValueTransformer.apply(
+                              mappedSinkRecord.getValue().detail(),
+                              removedContent -> put(removedContent, s3key),
+                              s3key)));
+
+      return success(mappedSinkRecord.getSinkRecord(), putEventsRequestEntry);
+    } catch (Exception e /*NoSuchAlgorithmException*/) {
+      return failure(mappedSinkRecord.getSinkRecord(), reportOnly("TODO", e));
+    }
+  }
+
+  private String generateS3Key(final SinkRecord sinkRecord) throws NoSuchAlgorithmException {
+    var payload = payloadOf(sinkRecord);
+    var s3Key = format("arn:aws:s3:::%s/%s", bucketName, toHex(sha256Of(payload)));
```

Comment:
@embano1 wdyt? The idea was to have a deterministic key for the generated S3 objects. In case of an error in the subsequent put of the Event Bridge even we do not need to upload the S3 objects again if we remember the already send key.

Reply:
> but I imagine a randomly generate hex

@baldawar do you mean purely random or deterministic random so we cover the requirement for idempotency outlined by @agebhar1 ?

Reply:
@embano1 you mentioned https://aws.amazon.com/about-aws/whats-new/2024/02/amazon-sqs-extended-client-library-python-payloads/ they use uuid4 as S3 key. We can internally map the SHA to the (uuid4 based) S3 key for de-duplication in case of an following error in put EventBridge event. The mapping must hold only the last 10 records, since one EventBridge event can only contain 10 items. The de-duplication can be optional to choose between possible duplicate S3 objects and speed (no SHA calculation. WDYT?

Reply:
> We can internally map the SHA to the (uuid4 based) S3 key

Good approach IMHO

> The mapping must hold only the last 10 records, since one EventBridge event can only contain 10 items

The only concern with this (and UUID) is if we're stuck in a crash loop (e.g., `PutEvents` fails and the connector is restarted by a watch daemon) we'd create a lot of S3 files. As long as we don't crash, keeping the last _n_ should be good (where _n_ is <=100 to not solely rely on today's `PutEvents` batch limits).

> The de-duplication can be optional to choose between possible duplicate S3 objects and speed (no SHA calculation. WDYT?

Premature optimization IMHO - let's go with SHA (or some suitable fingerprinting algorithm) for now unless we believe it's a huge performance bottleneck?

Reply:
> Question: How likely is duplicating S3 objects?

Depends on the scenario e.g., if we have a poison pill scenario (incl. insufficient perms writing to EventBridge, but granted write to S3) and the user sets up automatic connector/task restart we could get stuck creating (duplicating) S3 objects. If it's manual, i.e., operator is involved, I don't think it's that much of an issue (just couple duplicates), but again depends on the last offset checkpoint and how many events are consumed upon restart.

Reply:
I was implying purely random in my comment. I hadn't realized idempotency was a requirements (feel free to point me to links that should be reading upfront). 

For me idempotency based on SHA has been flaky in the past due to data corruption, packet bit-flips, or just plain old messy data from publishers. What has worked well in the past was explicitly making idempotency an opt-in feature and then asking publishers for a specific source of truth (like an "id" field within an event payload or a generator function like `<UUID> getIdempotencyToken(<Event>)`. 

FWIW; there's no reason to undo the current SHA based operations if you don't expect to change this part of the code, but if this is still WIP and expected to be refactored / improved; then using alternative options are worth considering.